reservation_units/utils/first_reservable_time_helper/first_reservable_time_helper.py

```diff
@@ -219,34 +269,85 @@ def calculate_all_first_reservable_times(self) -> None:
 
         if not AffectingTimeSpan.is_valid():
             AffectingTimeSpan.refresh()
-
-        chunk_size: int = qs.count()
-        if self.pagination_args is not None:
-            self.pagination_args["size"] = chunk_size
-            qs_slice = calculate_queryset_slice(**self.pagination_args)
-            chunk_size = qs_slice.stop - qs_slice.start
-            qs = qs[qs_slice.start :]
-
-        page_size: int = 0
-        # If we should only show reservable reservation units, increase the chunk size to max page size
-        # so that if filtering occurs, we don't need to fetch so many chunks.
-        size = graphene_settings.RELAY_CONNECTION_MAX_LIMIT if self.show_only_reservable else chunk_size
-
-        for reservation_unit in qs.hooked_iterator(self._get_elements_for_reservation_units, chunk_size=size):
+            cache.delete(self.cache_key)
+
+        has_valid_results_for_previous_pages = self._read_cached_results()
+        if has_valid_results_for_previous_pages:
+            # If we already have cached FRT results for enough reservation units to fill the page
+            # AND all the previous pages, then we don't need to calculate anything.
+            # NOTE: This does not support different orderings of the same queryset!
+            if len(self.first_reservable_times) >= self.stop_offset:
```

Comment:
Is the length check sufficient; i.e., there cannot be gaps in the data (but so that it extends to further pages because someone else had queried that far)?
I guess it at least may assume that the pages are queried in order from the API -- if so, I suppose we can make that assumption (at least for now because the frontend search page is the use case) -- unless the cache being shared between users might lead to some tricky situation I haven't thought through.

Reply:
It should be, since cached are unique per ordering too, and the entire cache gets invalidated an recalculated if even one of the items is no longer valid (i.e. items in page 1 become invalid -> must calculate all pages). Maybe there is some slight edge case when new reservation units are added, but those are very marginal cases that get corrected quickly.

Reply:
I was thinking about the case that if there's a page in the middle (say page 3) which has not been queried (but now page 4 is being queried) and its data has never been cached but enough later pages have been cached so that the length check passes. But I think this would only come to pass if someone(s) queries the pages out of sequence (e.g. 1, 2, 5, 4, 3) directly from the API, bypassing the frontend, and need not be supported for now.

Reply:
Yes, in regular use, you never query pages out of order. In any case, I don't think there is a way to know in that situation that the latter items are from page 4 and not 3 without calculating all of them.